### PR TITLE
Create user

### LINKED ISSUE #12: Authentication
#### As an application hoster I want to add user authentication to my application.

__Story__:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

Please split this task up into three separate PRs. The second, and third should
depend on the preceding ones.

__Acceptance criteria PR1__:
- [x] Users can sign up.

__Acceptance criteria PR2__:
- [ ] Users can log in after they've signed up.
- [ ] Users can log out after logging in.
- [ ] Logged in users should not be able to register, nor log in again.

__Acceptance criteria PR3__:
- [ ] Users can choose to be remembered as part of the login
  process.<sup>†</sup>
- [ ] Multiple browsers can concurrently maintain the remembered status for the
  same user.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser. Please be sure to completely
close the browser (`cmd` + `q`) when testing this functionality. Closing the
window (`cmd` + `w`, or clicking the red circle) on Mac OS does not close the
browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/modeling_users
- https://www.railstutorial.org/book/sign_up
- https://www.railstutorial.org/book/basic_login
- https://www.railstutorial.org/book/advanced_login

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by BinKuangHsiehApp**
Closes #12 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/models/user_test.rb [hunk 1]
```diff
@@ -1,7 +1,68 @@
...
+                           foo@bar_baz.com foo@bar+baz.com blackbvox@trijke..com]
+    invalid_addresses.each do |invalid_address|
+      @user.email = invalid_address
+      assert_not @user.valid?, "#{invalid_address.inspect} should be invalid"
+    end
+  end
+
+  test 'email addresses should be unique' do
+    duplicate_user = @user.dup
+    @user.save
```
Comment (bboe) [inline test/models/user_test.rb]: Always use `save!` if you aren't checking for the return value.

File: app/models/user.rb [hunk 2]
```diff
@@ -1,8 +1,7 @@
 class User < ApplicationRecord
   before_save {self.email = email.downcase}
   validates :name, presence: true
-  validates :email, presence: true
-  validates :email, format: {with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i}, uniqueness: {case_sensitive: false}
+  validates :email, presence: true, format: {with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i}, uniqueness: {case_sensitive: false}
```
Comment (bboe) [inline app/models/user.rb]: I see that you added presence validator here for email, but I don't see any associated tests in the commit.

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -1,5 +1,4 @@
 class ImagesController < ApplicationController
-
```
Comment (bboe) [inline app/controllers/images_controller.rb]: With regard to commit "add test for controller and user crud flow", tests should be added in the commit that tests the code. It makes the commits more atomic, and makes it easier to review (I always review commit by commit).

File: test/controllers/users_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  def test_new
+    get new_user_path
+    assert_response :ok
+  end
+
+  def test_create
+    assert_difference('User.count', 1) do
+      post users_path, params: user_params
+    end
+    assert_response :found
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Is there a flash message that should be asserted against?

File: test/controllers/users_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  def test_create__email_must_be_valid
+    assert_difference('User.count', 0) do
+      post users_path, params: user_params(email: 'invalid_email')
+    end
+    assert_select '.form-control-feedback', count: 1 do |feedback|
+      assert_equal 'is invalid', feedback.text
+    end
+    assert_response :ok
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Should this be an `ok` response when it fails?

File: test/controllers/users_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  def test_create
+    assert_difference('User.count', 1) do
+      post users_path, params: user_params
+    end
+    assert_response :found
+  end
+
+  def test_create__email_must_be_valid
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: These tests are redundant to the model tests. We need only test a single failure case for each set of possible failure cases as far as the controller is concerned.

File: test/models/user_test.rb [hunk 7]
```diff
@@ -19,17 +19,17 @@ def setup
...
     valid_addresses.each do |valid_address|
       @user.email = valid_address
       assert @user.valid?, "#{valid_address.inspect} should be valid"
     end
   end
 
   test 'email validation should reject invalid addresses' do
-    invalid_addresses = %w[user@example,com user_at_foo.org user.name@example.
-                           foo@bar_baz.com foo@bar+baz.com blackbvox@trijke..com]
+    invalid_addresses = %w(user@example,com user_at_foo.org user.name@example.
```
Comment (bboe) [inline test/models/user_test.rb]: It looks like you did some rubocop corrections here. Please make these changes in the commit that introduced the code.

File: db/migrate/20170828171710_create_users.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :name
```
[the remark this answers is not in the document]
Comment (BinKuangHsiehApp) [reply db/migrate/20170828171710_create_users.rb]: It should never be null. will address in migration file.

File: app/models/user.rb [hunk 9]
```diff
@@ -1,2 +1,8 @@
 class User < ApplicationRecord
+  before_save {self.email = email.downcase}
+  validates :name, presence: true
+  validates :email, presence: true
+  validates :email, format: {with: /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i}, uniqueness: {case_sensitive: false}
```
[the remark this answers is not in the document]
Comment (BinKuangHsiehApp) [reply app/models/user.rb]: use a gem and assume its been tested?

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  # test "the truth" do
```
[the remark this answers is not in the document]
Comment (BinKuangHsiehApp) [reply test/models/user_test.rb]: tests are added in the next commit. I squashed the two

File: test/flows/share_email_test.rb [hunk 11]
```diff
@@ -1,6 +1,6 @@
 require 'flow_test_helper'
 
-class ImagesCrudTest < FlowTestCase
+class ShareEmailTest < FlowTestCase
```
[the remark this answers is not in the document]
Comment (BinKuangHsiehApp) [reply test/flows/share_email_test.rb]: made it a separate commi

File: test/controllers/users_controller_test.rb (hunk 4 above)
Comment (BinKuangHsiehApp) [reply test/controllers/users_controller_test.rb]: yes just modified

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (BinKuangHsiehApp) [reply app/controllers/images_controller.rb]: I guess I should squash it!

File: test/controllers/users_controller_test.rb (hunk 5 above)
Comment (BinKuangHsiehApp) [reply test/controllers/users_controller_test.rb]: changed to unprocessable_entity

File: test/controllers/users_controller_test.rb (hunk 6 above)
Comment (BinKuangHsiehApp) [reply test/controllers/users_controller_test.rb]: removed the extra tests

File: test/controllers/users_controller_test.rb (hunk 4 above)
Comment (BinKuangHsiehApp) [reply test/controllers/users_controller_test.rb]: yes!

File: test/models/user_test.rb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                     password: 'foobar_andMore', password_confirmation: 'foobar_andMore')
+  end
+  test 'should be valid' do
+    assert_predicate @user, :valid?, "#{@user.inspect} should be valid."
+  end
+
+  test 'name should be present' do
+    @user.name = ' '
+    refute_predicate @user, :valid?, "#{@user.inspect} should not be valid if not present."
+    assert_equal "can't be blank", @user.errors[:name].first
```
Comment (bboe) [inline test/models/user_test.rb]: Could we compare to all the errors? In theory there should be only one (for all fields) as we want to isolate errors to specifically the one we're testing for.

File: test/models/user_test.rb [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  test 'email should be present not nil' do
```
Comment (bboe) [inline test/models/user_test.rb]: Glad to see nil being tested. Note that you can have multiple assertions for similar inputs in the same test case. Consider something like:

    [nil, '  '].each do |value|
      @user.name = value
      ..

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+      flash[:success] = 'You have successfully registered.'
+      redirect_to images_path
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  private
+
+  def _user_params
```
Comment (bboe) [inline app/controllers/users_controller.rb]: In ruby we don't often underscore prefix private methods. It doesn't hurt to, but it's not a common practice.

File: app/models/user.rb [hunk 15]
```diff
@@ -1,10 +1,10 @@
 class User < ApplicationRecord
-  before_save {self.email = email.downcase}
+  before_save { self.email = email.downcase }
```
Comment (bboe) [inline app/models/user.rb]: This change looks like it belongs in your second commit (where the code is added).

File: test/models/user_test.rb (hunk 13 above)
Comment (BinKuangHsiehApp) [reply test/models/user_test.rb]: nice!

File: test/models/user_test.rb [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    valid_address = 'user@example.com'
+    @user.email = valid_address
+    assert_predicate @user, :valid?, "#{valid_address.inspect} should be valid."
+  end
+
+  test 'email validation should reject invalid addresses' do
+    invalid_address = 'user@example,com'
+    @user.email = invalid_address
+    refute_predicate @user, :valid?, "#{invalid_address.inspect} should be invalid if email is invalid."
+    assert_equal ['is invalid'], @user.errors[:email]
```
Comment (bboe) [inline test/models/user_test.rb]: Looks like this assertion didn't get updated.

**Pull request merged**